wireless_modules/middle_module/test.py

```diff
@@ -0,0 +1,64 @@
+import paho.mqtt.client as mqtt
+import time
+import argparse
+
+
+parser = argparse.ArgumentParser(
+    description='MQTT wireless module test script with dummy payloads',
+    add_help=True)
+parser.add_argument(
+    '--module', action='store', type=int,
+    default=2, help="""The module number to determine the publish and subscription topics. 
+    If nothing is selected it will send on module number 2 topics.""")
+parser.add_argument(
+    '--host', action='store', type=str, default="5.196.95.208",
+    help="""Address of the MQTT broker. If nothing is selected it will
+    default to 5.196.95.208 - a test server by mosquito.""")
+
+
+if __name__ == "__main__":
+    client = mqtt.Client('my_id')
+    args = parser.parse_args()
+
+    MODULE_NUM = args.module
+
+    START_TOPIC = f'/v3/wireless-module/{MODULE_NUM}/start'
+    STOP_TOPIC = f'/v3/wireless-module/{MODULE_NUM}/stop'
+    SUB_DATA_TOPIC = f'/v3/wireless-module/{MODULE_NUM}/data'
+
+    # Broker domain name or IP address
+    BROKER = args.host
+
+    def on_connect(client, data, message):
+        print('Received: ', str(message.payload.decode('utf-8')))
+        return True
+
+    client.connect(BROKER)
+
+    # Calls the function on_connect whenever a message is received
+    client.on_message = on_connect
+    client.subscribe(SUB_DATA_TOPIC)
+
+    # The start loop starts scanning for incoming messages and the stop loop stops this process.
+    client.loop_start()
+    for i in range(1):
+        client.subscribe(SUB_DATA_TOPIC)
+        # Wait more than 10 seconds in case, main.py could not connect to the broker and takes 10 seconds to restart
+        time.sleep(11)
+        # Send repeatedly to make sure it's received
```

Comment:
Update: I tried using QoS 1 (QoS 2 is not supported by Micropython apparently) and in some instances it took me the third try to get the messages flowing through, this could be because the ESP32 wasn't checking for the message at the right time (but I assume QoS 1 would ensure that the packet is delivered so it would resend). So I would still suggest sending on the start topic multiple times (or sending again if we don't start receiving a message) just to make sure. And I don't really see any significant downside to doing so, unless I'm missing something here.

Reply:
QoS 1 is fine, cos as you said it's fine to receive the message more than once. Were you both publishing with QoS 1 and subscribing with QoS 1? As far as I know, not that I've really used it much, MQTT will use the minimum qos of the publish/subscribe for when the message goes from the broker to the subscribed client.

I agree, there's no significant downside to simply sending the message multiple times, but if we can get it to work,
- Using QoS would be neater/more "correct"
- Using QoS guarantees us that the message will be delivered, but with sending multiple times there is a chance that none of those will arrive at all.

It's actually quite strange that you're having reliability issues at all. I do wonder if something's not set up right, or if there's something different about the ESP that I'm not aware of.

Reply:
I'm using QoS 1 for both publishing and subcribing for the same reason you mention. 
When I talk about resending, we can resend with qos=1 (so I'm not suggesting remove using qos).
As for the set up and ESP32, I strongly believe the reason why it doesn't work sometimes is because the ESP32 sometimes fails to connect to the broker (I have seen this error appear a few times) and when this happens the ESP32 waits for 10 sec and reboots and then re-connnects and during this time the ESP32 is obviously not subscribed so even if we have qos=1 if the sender sends it before anyone subscribes, qos=1 doesn't help. 
However, this rebooting issues usually only occurs once and when you run it the second time it works, that's why I would suggest having a function which can check if the Mid module is sending any message and if not re-send the start message.

Reply:
I agree it sounds like a problem with the esp32 connecting to the broker. Rather than making a concession for this in the test code though, I think we really should focus on why the esp32 is failing to connect.

On the real bike, the `start` message will only be sent out once, and this test script needs to be able to ensure that the module is able to work consistently in this exact scenario. If the sensor module misses the start message, that's not the test script's problem, it has done it's job.